File: 01_Language/01_Functions/python/dirname.py
# coding: utf-8

import os


def dirname(path):
    end = len(path) - 1
    while end >= 1 and (path[end] == os.path.sep or path[end] == os.path.altsep):
        end -= 1
    path = path[0:end + 1]
    return os.path.dirname(path) or "."


if __name__ == '__main__':
    print(dirname("/etc/passwd"))
    print(dirname("/etc/"))
    print(dirname("/"))
    print(dirname("//"))
    print(dirname("."))
    print(dirname("C:\\"))
    print(dirname("/usr/local/lib"))

File: 01_Language/01_Functions/php/dirname.php
<?php

echo dirname("/etc/passwd") . PHP_EOL;
echo dirname("/etc/") . PHP_EOL;
echo dirname("/") . PHP_EOL;
echo dirname("//") . PHP_EOL;
echo dirname(".") . PHP_EOL;
echo dirname("C:\\") . PHP_EOL;
echo dirname("/usr/local/lib", 2);
